arpwatch.8
.\" @(#) $Id: arpwatch.8,v 1.13 2000/10/08 20:31:25 leres Exp $ (LBL)
.\"
.\" Copyright (c) 1992, 1994, 1996, 1997, 2000
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that: (1) source code distributions
.\" retain the above copyright notice and this paragraph in its entirety, (2)
.\" distributions including binary code include the above copyright notice and
.\" this paragraph in its entirety in the documentation or other materials
.\" provided with the distribution, and (3) all advertising materials mentioning
.\" features or use of this software display the following acknowledgement:
.\" ``This product includes software developed by the University of California,
.\" Lawrence Berkeley Laboratory and its contributors.'' Neither the name of
.\" the University nor the names of its contributors may be used to endorse
.\" or promote products derived from this software without specific prior
.\" written permission.
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
.\" WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
.\"
.TH ARPWATCH 8 "8 October 2000"
.UC 4
.SH NAME
arpwatch - keep track of ethernet/ip address pairings
.SH SYNOPSIS
.na
.B arpwatch
[
.B -dN
] [
.B -f
.I datafile
] [
.B -i
.I interface
]
.br
.ti +8
[
.B -n
.IR net [/ width
]] [
.B -r
.I file
]
.ad
.SH DESCRIPTION
.B Arpwatch
keeps track for ethernet/ip address pairings. It syslogs activity
and reports certain changes via email.
.B Arpwatch
uses
.BR pcap (3)
to listen for arp packets on a local ethernet interface.
.LP
The
.B -d
flag is used enable debugging. This also inhibits forking into the
background and emailing the reports. Instead, they are sent to
.IR stderr .
.LP
The
.B -f
flag is used to set the ethernet/ip address database filename.
The default is
.IR arp.dat .
.LP
The
.B -i
flag is used to override the default interface.
.LP
The
.B -n
flag specifies additional local networks. This can be useful to
avoid "bogon" warnings when there is more than one network running
on the same wire. If the optional
.I width
is not specified, the default netmask for the network's class is used.
.LP
The
.B -N
flag disables reporting any bogons.
.LP
The
.B -r
flag is used to specify a savefile
(perhaps created by
.BR tcpdump (1)
or
.BR pcapture (1))
to read from instead
of reading from the network. In this case,
.B arpwatch
does not fork.
.LP
Note that an empty
.I arp.dat
file must be created before the first time you run
.BR arpwatch .
.LP
.SH "REPORT MESSAGES"
Here's a quick list of the report messages generated by
.BR arpwatch (1)
(and
.BR arpsnmp (1)):
.TP
.B "new activity"
This ethernet/ip address pair has been used for the first time six
months or more.
.TP
.B "new station"
The ethernet address has not been seen before.
.TP
.B "flip flop"
The ethernet address has changed from the most recently seen address to
the second most recently seen address.
(If either the old or new ethernet address is a DECnet address and it
is less than 24 hours, the email version of the report is suppressed.)
.TP
.B "changed ethernet address"
The host switched to a new ethernet address.
.SH "SYSLOG MESSAGES"
Here are some of the syslog messages;
note that messages that are reported are also sysloged.
.TP
.B "ethernet broadcast"
The mac ethernet address of the host is a broadcast address.
.TP
.B "ip broadcast"
The ip address of the host is a broadcast address.
.TP
.B "bogon"
The source ip address is not local to the local subnet.
.TP
.B "ethernet broadcast"
The source mac or arp ethernet address was all ones or all zeros.
.TP
.B "ethernet mismatch"
The source mac ethernet address didn't match the address inside
the arp packet.
.TP
.B "reused old ethernet address"
The ethernet address has changed from the most recently seen address to
the third (or greater) least recently seen address.
(This is similar to a flip flop.)
.TP
.B "suppressed DECnet flip flop"
A "flip flop" report was suppressed because one of the two
addresses was a DECnet address.
.SH FILES
.na
.nh
.nf
/usr/operator/arpwatch - default directory
arp.dat - ethernet/ip address database
ethercodes.dat - vendor ethernet block list
.ad
.hy
.fi
.SH "SEE ALSO"
.na
.nh
.BR arpsnmp (8),
.BR arp (8),
.BR bpf (4),
.BR tcpdump (1),
.BR pcapture (1),
.BR pcap (3)
.ad
.hy
.SH AUTHORS
Craig Leres of the
Lawrence Berkeley National Laboratory Network Research Group,
University of California, Berkeley, CA.
.LP
The current version is available via anonymous ftp:
.LP
.RS
.I ftp://ftp.ee.lbl.gov/arpwatch.tar.gz
.RE
.SH BUGS
Please send bug reports to arpwatch@ee.lbl.gov.
.LP
Attempts are made to suppress DECnet flip flops but they
aren't always successful.
.LP
Most error messages are posted using syslog.


CHANGES
NG 1.6:
-Clean up the reporting via a central report function description table
o makes it easier to add custom reporting functions
-Add a paragraph to FAQ about compilation under cygwin
 thanks to Urs Meier for contributing
-Remove unneeded stuff
o massagevendor-old
o missingcodes.txt
o FILES
-Change char version[] to a const char *version for cleanliness
-Update arpsnmp a bit [which still is deprecated]
-Small cleanups here&there

NG 1.5:
-Report arp.dat and ethercode.dat reading failures in an explicit way
 thanks to Chuck Chargin Jr. for pointing it out

NG 1.4:
-Report all error arpwatch can detect via the standard report modes
 thanks to Russ Starr for suggesting
o New are
"bogon",
"ether broadcast",
"ether mismatch",
"ether too short",
"ether bad format",
"ether wrong type_ip",
"ether bad length",
"ether wrong op",
"ether wrong RARP",
"ether wrong type",
o this could mean breakage with old arpwatch operation!
-Make report mode 2 output of actions parseable by using "," as delimiter
-Remove ethercodes.dat.old - unused legacy
-Update ethercodes.dat from nmap-4.11
-Further static'fy local functions

NG 1.3:
-Buggy internal version only

NG 1.2:
-Fix make install target: man-pages will be installed now, too
 thanks to Aleksandr Shubnik
-Update ethercodes.dat from nmap-4.00

NG 1.1:
-Fix and update man-page to reflect the NG changes
-Add new feature: -P pcap_filter to use tcpdump-style filters
 thanks to John P. Rouillard for suggesting a filtering facility
-Improve error handling for pcap_compile() to print expression
-Small \n error-msg fix

NG 1.0:
-Fix small typo in man-page
-Allow getopt() to complain about unknown options (remove opterr=0)
-Rework help output and print to stdout, not stderr

DROP10:
-Fix longstanding and unnoticed BUG:
when running in default/daemon mode arp.dat was not checkpointed and as 
such never updated.

DROP9:
-Update ethercodes.dat from nmap-3.93
-Fix usage output; -t is for mailto, not -m (thanks to Jan Ostrochovsky)
-Add -F as mailfrom option (as everybody wants it; kudos Jan Ostrochovsky)
-Change CHECKPOINT (write arp DB to file) time to 3*60 seconds, was 10*60
-Change mkdep to also use mktemp if tempfile fails

DROP8:
-New configure option "--enable-fancy-mac" so MAC adresses are printed always
in xx:xx:.. notation e.g. 00:0a: instead of 0:0a

DROP7:
-Continue merging of Debian patches
o 11opt_sendmail_path: specify sendmail prog path
o 14opt_mailto: specify mailto address
o 15opt_username: drop privileges to this username (without restart)

-Disable savestr() - functionality covered by strndup()
-Make replacement strndup() safer with checks
-Drop into daemon mode the latest possible moment
o replace some syslog() calls with fprintf()s

DROP6:
-Merge a lot of vendor fixes from Debian (thanks to Peter Kelemen)
o 06debian_manpages: man-pages for arpfetch, arp2ethers
o 21arp2ethers: misc. awk fixes
o 26unconf_iface: continue if unconfigured interface, but don't use syslog()
o 25ignore_zero_ip: don't enter IP 0.0.0.0 into db
o 22secure_tempfile: use a secure tmpfile name in bihourly, mkdep and arpfetch
o 07debian_srcdir: add some missing $(srcdir)
o 12opt_nopromisc: -p not enter the device promisc mode
o 05debian_fhs: ethercodes into share/arpwatch, arp.dat into var/lib/arpwatch

-Add more fprintf(stderr, ...) on failures and make error reporting better
-Misc cleanups

DROP5:
-Remove #ifdef DEBUG altogether - enabled by default, so what.
 Since DROP2 I had no luck with stable releases, sorry.
-Add first & dirty make install hack to install necessary files
-Reorg and simplify arpwatch.c initialization code

DROP4.1:
-Some systems don't have strndup() - sorry, overread that it's a GNU extension
 so let configure check and have a replacement function ready [QUICKFIX]
 Thanks to Jim Sandoz

DROP4:
-Compile fix for arpwatch.c - report_mode was missing
-Segfault fix for arpwatch.c<-report.c: report_f was not initialized
-Apply vendor security patch against hostname overflow + strcpy()->strncpy()
-Shorten checkpoint time from 15 to 10 minutes (even lower values sensible?)

DROP3:
-Close stdin/out/err in daemon mode again
-report() is now a dispatching function for the actual reporting function
-Change notification codes from clear-text("new station") to int
-Add defines for the former and lookup table
-Change all calls to report() to use ACTION_ defines
-DECNet Flipflops are also reported via report(), not dosyslog()

DROP2:
-Add stdout reporting mode (switch -m)
-Update ethercodes from nmap-3.81

DROP1:
-Remove all unnecessary (void) casts when retcode was not used
-Update ethercodes.dat to the ones from nmap
-indent the code
-Remove register hints for variables (nothing really considered)
-Remove {scss,rcs}id 


-- FORKED Mon Feb  7 13:31:36 CET 2005

@(#) $Id: CHANGES,v 1.41 2004/01/22 22:25:43 leres Exp $ (LBL)

v2.1 Thu Jan 22 14:05:27 PST 2004

- Add FDDI support. Based on code contributed by Flemming Johansen
  (fsj@csd.cri.dk)

- New -f flag overrides the default database filename.

- New -r flag allows tcpdump save files to be read.

- memcpy() doesn't handle overlapping buffers so use bcopy() instead.
  Reported by Attila Bruncsak (attila.bruncsak@itu.int)

- Have configure determine the sendmail path.

- Don't bother updating hostnames when reading arp.dat since this can
  be expensive in the number of dns requests generated.

- Improve robostness of the aprfetch script.

- Update massagevendor to convert the IEEE public Organizationally
  Unique Identifier (OUI) file format.

- Don't use -e with awk, not all versions support it. Suggested by
  Stephen Thompson (thompson@pobox.upenn.edu)

- Convert usage of mktemp() to mkstemp().

- Add -n flag (alternate networks) and -N flag (don't report bogons).

- Improve ethernet address parser error checking.

- More concessions to linux.

- Special case dn_skipname() check for bind 8.

- Convert arp2ethers from csh to sh.

v2.0.2 Sat Jun  7 03:15:03 PDT 1997


arpwatch.c
/*
 * Copyright (c) 1990, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 2000
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that: (1) source code distributions
 * retain the above copyright notice and this paragraph in its entirety, (2)
 * distributions including binary code include the above copyright notice and
 * this paragraph in its entirety in the documentation or other materials
 * provided with the distribution, and (3) all advertising materials mentioning
 * features or use of this software display the following acknowledgement:
 * ``This product includes software developed by the University of California,
 * Lawrence Berkeley Laboratory and its contributors.'' Neither the name of
 * the University nor the names of its contributors may be used to endorse
 * or promote products derived from this software without specific prior
 * written permission.
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */
#ifndef lint
static const char copyright[] =
    "Copyright (c) 1990, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 2000\n\
The Regents of the University of California.  All rights reserved.\n";
static const char rcsid[] =
    "@(#) $Id: arpwatch.c,v 1.63 2000/10/14 02:07:28 leres Exp $ (LBL)";
#endif

/*
 * arpwatch - keep track of ethernet/ip address pairings, report changes
 */

#include <sys/param.h>
#include <sys/types.h>				/* concession to AIX */
#include <sys/file.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <sys/time.h>

#if __STDC__
struct mbuf;
struct rtentry;
#endif
#include <net/if.h>

#include <netinet/in.h>
#include <netinet/if_ether.h>

#include <arpa/inet.h>

#include <ctype.h>
#include <errno.h>
#ifdef HAVE_MEMORY_H
#include <memory.h>
#endif
#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#endif
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <unistd.h>

#include <pcap.h>

#include "gnuc.h"
#ifdef HAVE_OS_PROTO_H
#include "os-proto.h"
#endif

#include "arpwatch.h"
#include "db.h"
#include "ec.h"
#include "fddi.h"
#include "file.h"
#include "machdep.h"
#include "setsignal.h"
#include "util.h"

/* Some systems don't define these */
#ifndef ETHERTYPE_REVARP
#define ETHERTYPE_REVARP	0x8035
#endif

#ifndef REVARP_REQUEST
#define REVARP_REQUEST		3
#define REVARP_REPLY		4
#endif

#ifndef ETHERTYPE_TRAIL
#define ETHERTYPE_TRAIL		0x1000
#endif

#ifndef ETHERTYPE_APOLLO
#define ETHERTYPE_APOLLO	0x8019
#endif

#ifndef IN_CLASSD_NET
#define IN_CLASSD_NET		0xf0000000
#endif

#ifndef max
#define max(a,b) ((b)>(a)?(b):(a))
#endif

char *prog;

int can_checkpoint;
int swapped;
int nobogons;

static u_int32_t net;
static u_int32_t netmask;

struct nets {
	u_int32_t net;
	u_int32_t netmask;
};

static struct nets *nets;
static int nets_ind;
static int nets_size;

extern int optind;
extern int opterr;
extern char *optarg;

/* Forwards */
int	addnet(const char *);
RETSIGTYPE checkpoint(int);
RETSIGTYPE die(int);
int	isbogon(u_int32_t);
int	main(int, char **);
void	process_ether(u_char *, const struct pcap_pkthdr *, const u_char *);
void	process_fddi(u_char *, const struct pcap_pkthdr *, const u_char *);
int	readsnmp(char *);
int	snmp_add(u_int32_t, u_char *, time_t, char *);
int	sanity_ether(struct ether_header *, struct ether_arp *, int);
int	sanity_fddi(struct fddi_header *, struct ether_arp *, int);
__dead	void usage(void) __attribute__((volatile));

int
main(int argc, char **argv)
{
	register char *cp;
	register int op, pid, snaplen, timeout, linktype, status;
#ifdef TIOCNOTTY
	register int fd;
#endif
	register pcap_t *pd;
	register char *interface, *rfilename;
	struct bpf_program code;
	char errbuf[PCAP_ERRBUF_SIZE];
	char options[] =
		"d"
		/**/
		/**/
		"f:"
		/**/
		/**/
		"i:"
		/**/
		/**/
		"n:"
		/**/
		/**/
		"N"
		/**/
		/**/
		"r:"
		/**/
		/**/
	;

	if (argv[0] == NULL)
		prog = "arpwatch";
	else if ((cp = strrchr(argv[0], '/')) != NULL)
		prog = cp + 1;
	else
		prog = argv[0];

	if (abort_on_misalignment(errbuf) < 0) {
		(void)fprintf(stderr, "%s: %s\n", prog, errbuf);
		exit(1);
	}

	opterr = 0;
	interface = NULL;
	rfilename = NULL;
	pd = NULL;
	while ((op = getopt(argc, argv, options)) != EOF)
		switch (op) {

		case 'd':
			++debug;
#ifndef DEBUG
			(void)fprintf(stderr,
			    "%s: Warning: Not compiled with -DDEBUG\n", prog);
#endif
			break;

		case 'f':
			arpfile = optarg;
			break;

		case 'i':
			interface = optarg;
			break;

		case 'n':
			if (!addnet(optarg))
				usage();
			break;

		case 'N':
			++nobogons;
			break;

		case 'r':
			rfilename = optarg;
			break;
		/**/
		/**/
		default:
			usage();
		}

	if (optind != argc)
		usage();

	if (rfilename != NULL) {
		net = 0;
		netmask = 0;
	} else {
		/* Determine interface if not specified */
		if (interface == NULL &&
		    (interface = pcap_lookupdev(errbuf)) == NULL) {
			(void)fprintf(stderr, "%s: lookup_device: %s\n",
			    prog, errbuf);
			exit(1);
		}

		/* Determine network and netmask */
		if (pcap_lookupnet(interface, &net, &netmask, errbuf) < 0) {
			(void)fprintf(stderr, "%s: bad interface %s: %s\n",
			    prog, interface, errbuf);
			exit(1);
		}

		/* Drop into the background if not debugging */
		if (!debug) {
			pid = fork();
			if (pid < 0) {
				syslog(LOG_ERR, "main fork(): %m");
				exit(1);
			} else if (pid != 0)
				exit(0);
			(void)close(fileno(stdin));
			(void)close(fileno(stdout));
			(void)close(fileno(stderr));
#ifdef TIOCNOTTY
			fd = open("/dev/tty", O_RDWR);
			if (fd >= 0) {
				(void)ioctl(fd, TIOCNOTTY, 0);
				(void)close(fd);
			}
#else
			(void) setsid();
#endif
		}
	}

	openlog(prog, 0, LOG_DAEMON);

	if (chdir(arpdir) < 0) {
		syslog(LOG_ERR, "chdir(%s): %m", arpdir);
		syslog(LOG_ERR, "(using current working directory)");
	}

	if (rfilename != NULL) {
		pd = pcap_open_offline(rfilename, errbuf);
		if (pd == NULL) {
			syslog(LOG_ERR, "pcap open %s: %s", rfilename,  errbuf);
			exit(1);
		}
		swapped = pcap_is_swapped(pd);
	} else {
		snaplen = max(sizeof(struct ether_header),
		    sizeof(struct fddi_header)) + sizeof(struct ether_arp);
		timeout = 1000;
		pd = pcap_open_live(interface, snaplen, 1, timeout, errbuf);
		if (pd == NULL) {
			syslog(LOG_ERR, "pcap open %s: %s", interface, errbuf);
			exit(1);
		}
#ifdef WORDS_BIGENDIAN
		swapped = 1;
#endif
	}

	/*
	 * Revert to non-privileged user after opening sockets
	 * (not needed on most systems).
	 */
	setgid(getgid());
	setuid(getuid());

	/* Must be ethernet or fddi */
	linktype = pcap_datalink(pd);
	if (linktype != DLT_EN10MB && linktype != DLT_FDDI) {
		syslog(LOG_ERR, "Link layer type %d not ethernet or fddi",
		    linktype);
		exit(1);
	}

	/* Compile and install filter */
	if (pcap_compile(pd, &code, "arp or rarp", 1, netmask) < 0) {
		syslog(LOG_ERR, "pcap_compile: %s", pcap_geterr(pd));
		exit(1);
	}
	if (pcap_setfilter(pd, &code) < 0) {
		syslog(LOG_ERR, "pcap_setfilter: %s", pcap_geterr(pd));
		exit(1);
	}
	if (rfilename == NULL)
		syslog(LOG_INFO, "listening on %s", interface);

	/* Read in database */
	initializing = 1;
	if (!readdata())
		exit(1);
	sorteinfo();
#ifdef DEBUG
	if (debug > 2) {
		debugdump();
		exit(0);
	}
#endif
	initializing = 0;

	(void)setsignal(SIGINT, die);
	(void)setsignal(SIGTERM, die);
	(void)setsignal(SIGHUP, die);
	if (rfilename == NULL) {
		(void)setsignal(SIGQUIT, checkpoint);
		(void)setsignal(SIGALRM, checkpoint);
		(void)alarm(CHECKPOINT);
	}

	switch (linktype) {

	case DLT_EN10MB:
		status = pcap_loop(pd, 0, process_ether, NULL);
		break;

	case DLT_FDDI:
		status = pcap_loop(pd, 0, process_fddi, NULL);
		break;

	default:
		syslog(LOG_ERR, "bad linktype %d (can't happen)", linktype);
		exit(1);
	}
	if (status < 0) {
		syslog(LOG_ERR, "pcap_loop: %s", pcap_geterr(pd));
		exit(1);
	}
	pcap_close(pd);
	if (!dump())
		exit(1);
	exit(0);
}

/* Process an ethernet arp/rarp packet */
void
process_ether(register u_char *u, register const struct pcap_pkthdr *h,
    register const u_char *p)
{
	register struct ether_header *eh;
	register struct ether_arp *ea;
	register u_char *sea, *sha;
	register time_t t;
	u_int32_t sia;

	eh = (struct ether_header *)p;
	ea = (struct ether_arp *)(eh + 1);

	if (!sanity_ether(eh, ea, h->caplen))
		return;

	/* Source hardware ethernet address */
	sea = (u_char *)ESRC(eh);

	/* Source ethernet address */
	sha = (u_char *)SHA(ea);

	/* Source ip address */
	BCOPY(SPA(ea), &sia, 4);

	/* Watch for bogons */
	if (isbogon(sia)) {
		dosyslog(LOG_INFO, "bogon", sia, sea, sha);
		return;
	}

	/* Watch for ethernet broadcast */
	if (MEMCMP(sea, zero, 6) == 0 || MEMCMP(sea, allones, 6) == 0 ||
	    MEMCMP(sha, zero, 6) == 0 || MEMCMP(sha, allones, 6) == 0) {
		dosyslog(LOG_INFO, "ethernet broadcast", sia, sea, sha);
		return;
	}

	/* Double check ethernet addresses */
	if (MEMCMP(sea, sha, 6) != 0) {
		dosyslog(LOG_INFO, "ethernet mismatch", sia, sea, sha);
		return;
	}

	/* Got a live one */
	t = h->ts.tv_sec;
	can_checkpoint = 0;
	if (!ent_add(sia, sea, t, NULL))
		syslog(LOG_ERR, "ent_add(%s, %s, %ld) failed",
		    intoa(sia), e2str(sea), t);
	can_checkpoint = 1;
}

/* Perform sanity checks on an ethernet arp/rarp packet, return true if ok */
int
sanity_ether(register struct ether_header *eh, register struct ether_arp *ea,
    register int len)
{
	/* XXX use bsd style ether_header to avoid messy ifdef's */
	struct bsd_ether_header {
		u_char  ether_dhost[6];
		u_char  ether_shost[6];
		u_short ether_type;
	};
	register u_char *shost = ((struct bsd_ether_header *)eh)->ether_shost;

	eh->ether_type = ntohs(eh->ether_type);
	ea->arp_hrd = ntohs(ea->arp_hrd);
	ea->arp_pro = ntohs(ea->arp_pro);
	ea->arp_op = ntohs(ea->arp_op);

	if (len < sizeof(*eh) + sizeof(*ea)) {
		syslog(LOG_ERR, "short (want %d)\n", sizeof(*eh) + sizeof(*ea));
		return(0);
	}

	/* XXX sysv r4 seems to use hardware format 6 */
	if (ea->arp_hrd != ARPHRD_ETHER && ea->arp_hrd != 6) {
		syslog(LOG_ERR, "%s sent bad hardware format 0x%x\n",
		    e2str(shost), ea->arp_hrd);
		return(0);
	}

	/* XXX hds X terminals sometimes send trailer arp replies */
	if (ea->arp_pro != ETHERTYPE_IP && ea->arp_pro != ETHERTYPE_TRAIL) {
		syslog(LOG_ERR, "%s sent packet not ETHERTYPE_IP (0x%x)\n",
		    e2str(shost), ea->arp_pro);
		return(0);
	}

	if (ea->arp_hln != 6 || ea->arp_pln != 4) {
		syslog(LOG_ERR, "%s sent bad addr len (hard %d, prot %d)\n",
		    e2str(shost), ea->arp_hln, ea->arp_pln);
		return(0);
	}

	/*
	 * We're only interested in arp requests, arp replies
	 * and reverse arp replies
	 */
	if (eh->ether_type == ETHERTYPE_ARP) {
		if (ea->arp_op != ARPOP_REQUEST &&
		    ea->arp_op != ARPOP_REPLY) {
			syslog(LOG_ERR, "%s sent wrong arp op %d\n",
			     e2str(shost), ea->arp_op);
			return(0);
		}
	} else if (eh->ether_type == ETHERTYPE_REVARP) {
		if (ea->arp_op == REVARP_REQUEST) {
			/* no useful information here */
			return(0);
		} else if (ea->arp_op != REVARP_REPLY) {
			if (debug)
				syslog(LOG_ERR, "%s sent wrong revarp op %d\n",
				    e2str(shost), ea->arp_op);
			return(0);
		}
	} else {
		syslog(LOG_ERR, "%s sent bad type 0x%x\n",
		    e2str(shost), eh->ether_type);
		return(0);
	}

	return(1);
}

static void
bit_reverse(u_char *p, unsigned len)
{
	unsigned i;
	u_char b;

	for (i = len; i; i--, p++) {
		b = (*p & 0x01 ? 0x80 : 0)
			| (*p & 0x02 ? 0x40 : 0)
			| (*p & 0x04 ? 0x20 : 0)
			| (*p & 0x08 ? 0x10 : 0)
			| (*p & 0x10 ? 0x08 : 0)
			| (*p & 0x20 ? 0x04 : 0)
			| (*p & 0x40 ? 0x02 : 0)
			| (*p & 0x80 ? 0x01 : 0);
		*p = b;
	}
}

void
process_fddi(register u_char *u, register const struct pcap_pkthdr *h,
    register const u_char *p)
{
	register struct fddi_header *fh;
	register struct ether_arp *ea;
	register u_char *sea, *sha;
	register time_t t;
	u_int32_t sia;

	fh = (struct fddi_header *)p;
	ea = (struct ether_arp *)(fh + 1);

	if (!swapped) {
		bit_reverse(fh->src, 6);
		bit_reverse(fh->dst, 6);
	}
	if (!sanity_fddi(fh, ea, h->caplen))
		return;

	/* Source MAC hardware ethernet address */
	sea = (u_char *)fh->src;

	/* Source ARP ethernet address */
	sha = (u_char *)SHA(ea);

	/* Source ARP ip address */
	BCOPY(SPA(ea), &sia, 4);

	/* Watch for bogons */
	if (isbogon(sia)) {
		dosyslog(LOG_INFO, "bogon", sia, sea, sha);
		return;
	}

	/* Watch for ethernet broadcast */
	if (MEMCMP(sea, zero, 6) == 0 || MEMCMP(sea, allones, 6) == 0 ||
	    MEMCMP(sha, zero, 6) == 0 || MEMCMP(sha, allones, 6) == 0) {
		dosyslog(LOG_INFO, "ethernet broadcast", sia, sea, sha);
		return;
	}

	/* Double check ethernet addresses */
	if (MEMCMP(sea, sha, 6) != 0) {
		dosyslog(LOG_INFO, "ethernet mismatch", sia, sea, sha);
		return;
	}

	/* Got a live one */
	t = h->ts.tv_sec;
	can_checkpoint = 0;
	if (!ent_add(sia, sea, t, NULL))
		syslog(LOG_ERR, "ent_add(%s, %s, %ld) failed",
		    intoa(sia), e2str(sea), t);
	can_checkpoint = 1;
}

/* Perform sanity checks on arp/rarp packet, return true if ok */
int
sanity_fddi(register struct fddi_header *fh, register struct ether_arp *ea,
    register int len)
{
	u_char *shost = fh->src;
	u_short type, hrd, pro, op;

	/* This rather clunky copy stuff is needed because the fddi header
	 * has an odd (i.e. not even) length, causing memory alignment
	 * errors when attempts are made to access the arp header fields
	 * as shorts */
	BCOPY(fh->snap.snap_type, &type, sizeof(u_short));
	BCOPY(&(ea->arp_hrd), &hrd, sizeof(hrd));
	BCOPY(&(ea->arp_pro), &pro, sizeof(pro));
	BCOPY(&(ea->arp_op), &op, sizeof(op));
	type = ntohs(type);
	hrd = ntohs(hrd);
	pro = ntohs(pro);
	op = ntohs(op);

	if (len < sizeof(*fh) + sizeof(*ea)) {
		syslog(LOG_ERR, "short (want %d)\n", sizeof(*fh) + sizeof(*ea));
		return(0);
	}

	/* XXX sysv r4 seems to use hardware format 6 */
	if (hrd != ARPHRD_ETHER && hrd != 6) {
		syslog(LOG_ERR, "%s sent bad hardware format 0x%x\n",
		    e2str(shost), hrd);
		return(0);
	}


	/* XXX hds X terminals sometimes send trailer arp replies */
	if (pro != ETHERTYPE_IP &&
	    pro != ETHERTYPE_TRAIL &&
	    pro != ETHERTYPE_APOLLO) {
		syslog(LOG_ERR, "%s sent packet not ETHERTYPE_IP (0x%x)\n",
		    e2str(shost), pro);
		return(0);
	}

	if (ea->arp_hln != 6 || ea->arp_pln != 4) {
		syslog(LOG_ERR, "%s sent bad addr len (hard %d, prot %d)\n",
		    e2str(shost), ea->arp_hln, ea->arp_pln);
		return(0);
	}

	/*
	 * We're only interested in arp requests, arp replies
	 * and reverse arp replies
	 */
	if (type == ETHERTYPE_ARP) {
		if (op != ARPOP_REQUEST &&
		    op != ARPOP_REPLY) {
			syslog(LOG_ERR, "%s sent wrong arp op %d\n",
			     e2str(shost), op);
			return(0);
		}
	} else if (type == ETHERTYPE_REVARP) {
		if (op == REVARP_REQUEST) {
			/* no useful information here */
			return(0);
		} else if (op != REVARP_REPLY) {
			if (debug)
				syslog(LOG_ERR, "%s sent wrong revarp op %d\n",
				    e2str(shost), op);
			return(0);
		}
	} else {
		syslog(LOG_ERR, "%s sent bad type 0x%x\n",
		    e2str(shost), type);
		return(0);
	}
	return(1);
}

int
addnet(register const char *str)
{
	register char *cp;
	register int width;
	register u_int32_t n, m;
	register struct nets *np;
	char *cp2;
	char tstr[64];

	if (strlen(str) > sizeof(tstr) - 1)
		return(0);

	if (nets_size <= 0) {
		nets_size = 8;
		nets = malloc(nets_size * sizeof(*nets));
	} else if (nets_size <= nets_ind) {
		/* XXX debugging */
		nets_size <<= 1;
		nets = realloc(nets, nets_size * sizeof(*nets));
	}
	if (nets == NULL) {
		(void)fprintf(stderr, "%s: addnet: malloc/realloc: %s\n",
		    prog, strerror(errno));
		exit(1);
	}
	np = nets + nets_ind;

	width = 0;
	strcpy(tstr, str);
	cp = strchr(tstr, '/');
	if (cp != NULL) {
		*cp++ = '\0';
		width = strtol(cp, &cp2, 10);
		/* Trailing garbage */
		if (*cp2 != '\0')
			    return (0);
		if (width > 32)
			    return (0);
	}

	/* XXX hack */
	n = ntohl(inet_addr(tstr));
	while ((n & 0xff000000) == 0) {
		n <<= 8;
		if (n == 0)
			return (0);
	}
	n = htonl(n);

	if (width != 0) {
		m = ~0;
		m <<= 32 - width;
	} else if (IN_CLASSA(n))
		m = IN_CLASSA_NET;
	else if (IN_CLASSB(n))
		m = IN_CLASSB_NET;
	else if (IN_CLASSC(n))
		m = IN_CLASSC_NET;
	else if (IN_CLASSD(n))
		m = IN_CLASSD_NET;
	else
		return (0);
	m = htonl(m);

	np->net = n;
	np->netmask = m;
	++nets_ind;

	return (1);
}

int
isbogon(register u_int32_t sia)
{
	register int i;
	register struct nets *np;

	if (nobogons)
		return (0);
	if ((sia & netmask) == net)
		return (0);
	for (i = 0, np = nets; i < nets_ind; ++i, ++np)
		if ((sia & np->netmask) == np->net)
			return (0);
	return (1);
}

RETSIGTYPE
die(int signo)
{

	syslog(LOG_DEBUG, "exiting");
	checkpoint(0);
	exit(1);
}

RETSIGTYPE
checkpoint(int signo)
{

	if (!can_checkpoint)
		(void)alarm(1);
	else {
		(void)alarm(0);
		(void)dump();
		(void)alarm(CHECKPOINT);
	}
	return RETSIGVAL;
}

__dead void
usage(void)
{
	extern char version[];
	char usage[] =
		"[-dN] "
		/**/
		/**/
		"[-f datafile] "
		/**/
		/**/
		"[-i interface] "
		/**/
		/**/
		"[-n net[/width]] "
		/**/
		/**/
		"[-r file] "
		/**/
		/**/
		"\n"
	;

	(void)fprintf(stderr, "Version %s\n", version);
	(void)fprintf(stderr, "usage: %s %s", prog, usage);
	exit(1);
}
